les/vflux/server/clientpool.go

```diff
@@ -0,0 +1,375 @@
+// priority and remains inactive for longer than the allowed timeout then it will be
+// disconnected by calling the Disconnect function of the clientPeer interface.
+func (cp *ClientPool) Register(peer clientPeer) ConnectedBalance {
+	cp.ns.SetField(peer.Node(), clientField, clientPeerInstance{peer})
+	balance, _ := cp.ns.GetField(peer.Node(), btSetup.balanceField).(*nodeBalance)
+	return balance
+}
+
+// Unregister removes the peer from the client pool
+func (cp *ClientPool) Unregister(peer clientPeer) {
+	cp.ns.SetField(peer.Node(), clientField, nil)
+}
+
+// SetDefaultFactors sets the default price factors applied to subsequently connected clients
+func (cp *ClientPool) SetDefaultFactors(posFactors, negFactors PriceFactors) {
+	cp.lock.Lock()
+	cp.defaultPosFactors = posFactors
+	cp.defaultNegFactors = negFactors
+	cp.lock.Unlock()
+}
+
+// setConnectedBias sets the connection bias, which is applied to already connected clients
+// So that already connected client won't be kicked out very soon and we can ensure all
+// connected clients can have enough time to request or sync some data.
+func (cp *ClientPool) SetConnectedBias(bias time.Duration) {
+	cp.lock.Lock()
+	cp.connectedBias = bias
+	cp.setActiveBias(bias)
+	cp.lock.Unlock()
+}
+
+// SetCapacity sets the assigned capacity of a connected client
+func (cp *ClientPool) SetCapacity(node *enode.Node, reqCap uint64, bias time.Duration, requested bool) (capacity uint64, err error) {
+	cp.lock.RLock()
+	if cp.connectedBias > bias {
+		bias = cp.connectedBias
+	}
+	cp.lock.RUnlock()
+
+	cp.ns.Operation(func() {
+		balance, _ := cp.ns.GetField(node, btSetup.balanceField).(*nodeBalance)
+		if balance == nil {
+			err = ErrNotConnected
+			return
+		}
+		capacity, _ = cp.ns.GetField(node, ppSetup.capacityField).(uint64)
+		if capacity == 0 {
+			// if the client is inactive then it has insufficient priority for the minimal capacity
+			// (will be activated automatically with minCap when possible)
+			return
+		}
+		if reqCap < cp.minCap {
+			// can't request less than minCap; switching between 0 (inactive state) and minCap is
+			// performed by the server automatically as soon as necessary/possible
+			reqCap = cp.minCap
+		}
+		if reqCap > cp.minCap {
+			if cp.ns.GetState(node).HasNone(btSetup.priorityFlag) && reqCap > cp.minCap {
+				err = ErrNoPriority
+				return
+			}
+		}
+		if reqCap == capacity {
+			return
+		}
+		curveBias := bias
+		if requested {
+			// mark the requested node so that the UpdateCapacity callback can signal
+			// whether the update is the direct result of a SetCapacity call on the given node
+			cp.capReqNode = node
+			defer func() {
+				cp.capReqNode = nil
+			}()
+		}
+
+		// estimate maximum available capacity at the current priority level and request
+		// the estimated amount; allow a limited number of retries because individual
+		// balances can change between the estimation and the request
+		for count := 0; count < 20; count++ {
```

Comment:
I'm not sure the retry is necessary. Serving this function should be quick enough.

Reply:
I did not really like this part either so I changed it a bit. Now I think it's much nicer, I also added some explanatory comments. So basically it's the `priorityPool` that determines the final capacity (as it should be), `enforceLimits` adjusts the requesting node in finer steps than the rest of the nodes but otherwise it's the same mechanism. The capacity curve estimation is just an optimization layer on top of it, it prevents requesting too high capacity because that could cost a lot of iterations in `enforceLimits`. The estimation itself is very quick and it guarantees that `enforceLimits` will either find the best possible capacity in 5 iterations or fail if there is something wrong with the estimation (this case is detected by the fuzzer).